### ISSUE TITLE
"msg": "Authentication failed: Request failed: '_io.BytesIO' object has no attribute 'getbuffer'"

### ISSUE THREAD

**Issue by rg5147** [CiscoDevNet/ansible-mso#336]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description
When executing a specific playbook after updating the MSO collection, I am seeing the following error for this part of the playbook.
This is only seen for user that had an upgraded collection. I can't pinpoint which module is the cause of this error and why.

```
##########################################
#        TEST TASK TO CHECK ND VERSION
##########################################

    - name: Get MSO version from MSO >= 3.2
      cisco.mso.mso_version:
        <<: *mso_login
        state: query
        output_level: debug
```
```
TASK [Get MSO version from MSO >= 3.2] ******************************************************************************************************************************************
task path: /usr/abc4/playbooks/mso-adv-bootcamp_undeploy_schema_template_tenant_abc4_only_test.yml:43
<10.122.186.69> attempting to start connection
<10.122.186.69> using connection plugin ansible.netcommon.httpapi
<10.122.186.69> local domain socket does not exist, starting it
<10.122.186.69> control socket path is /root/.ansible/pc/59162f5832
<10.122.186.69> local domain socket listeners started successfully
<10.122.186.69> loaded API plugin ansible_collections.cisco.nd.plugins.httpapi.nd from path /root/.ansible/collections/ansible_collections/cisco/nd/plugins/httpapi/nd.py for platform type cisco.nd.nd
<10.122.186.69> 
<10.122.186.69> local domain socket path is /root/.ansible/pc/59162f5832
<10.122.186.69> ESTABLISH LOCAL CONNECTION FOR USER: root
<10.122.186.69> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp/ansible-local-62998I1X_g `"&& mkdir "` echo /root/.ansible/tmp/ansible-local-62998I1X_g/ansible-tmp-1677720599.45-6308-75782741894533 `" && echo ansible-tmp-1677720599.45-6308-75782741894533="` echo /root/.ansible/tmp/ansible-local-62998I1X_g/ansible-tmp-1677720599.45-6308-75782741894533 `" ) && sleep 0'
Using module_utils file ansible_collections/cisco/mso/plugins/module_utils
Using module_utils file ansible_collections/cisco/mso/plugins/module_utils/mso
Using module_utils file ansible_collections/cisco/mso
Using module_utils file ansible_collections
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/basic.py
Using module_utils file ansible_collections/cisco/mso/plugins
Using module_utils file ansible_collections/cisco
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/six/__init__.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/urls.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/connection.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/_text.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/common/json.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/common/__init__.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/common/_collections_compat.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/common/collections.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/common/text/formatters.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/common/validation.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/common/text/converters.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/pycompat24.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/common/text/__init__.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/common/process.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/common/_json_compat.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/common/_utils.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/common/parameters.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/parsing/convert_bool.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/parsing/__init__.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/compat/selectors.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/common/sys_info.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/compat/__init__.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/common/file.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/compat/_selectors2.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/distro/__init__.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/distro/_distro.py
<10.122.186.69> Attempting python interpreter discovery
<10.122.186.69> EXEC /bin/sh -c 'echo PLATFORM; uname; echo FOUND; command -v '"'"'/usr/bin/python'"'"'; command -v '"'"'python3.7'"'"'; command -v '"'"'python3.6'"'"'; command -v '"'"'python3.5'"'"'; command -v '"'"'python2.7'"'"'; command -v '"'"'python2.6'"'"'; command -v '"'"'/usr/libexec/platform-python'"'"'; command -v '"'"'/usr/bin/python3'"'"'; command -v '"'"'python'"'"'; echo ENDFOUND && sleep 0'
<10.122.186.69> EXEC /bin/sh -c '/usr/bin/python && sleep 0'
Using module file /root/.ansible/collections/ansible_collections/cisco/mso/plugins/modules/mso_version.py
<10.122.186.69> PUT /root/.ansible/tmp/ansible-local-62998I1X_g/tmp9MYLdp TO /root/.ansible/tmp/ansible-local-62998I1X_g/ansible-tmp-1677720599.45-6308-75782741894533/AnsiballZ_mso_version.py
<10.122.186.69> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-local-62998I1X_g/ansible-tmp-1677720599.45-6308-75782741894533/ /root/.ansible/tmp/ansible-local-62998I1X_g/ansible-tmp-1677720599.45-6308-75782741894533/AnsiballZ_mso_version.py && sleep 0'
<10.122.186.69> EXEC /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-local-62998I1X_g/ansible-tmp-1677720599.45-6308-75782741894533/AnsiballZ_mso_version.py && sleep 0'
<10.122.186.69> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-local-62998I1X_g/ansible-tmp-1677720599.45-6308-75782741894533/ > /dev/null 2>&1 && sleep 0'
The full traceback is:
WARNING: The below traceback may *not* be related to the actual failure.
  File "/tmp/ansible_cisco.mso.mso_version_payload_RbKPgq/ansible_cisco.mso.mso_version_payload.zip/ansible_collections/cisco/mso/plugins/module_utils/mso.py", line 557, in request
    info = self.connection.send_request(method, uri, json.dumps(data))
  File "/tmp/ansible_cisco.mso.mso_version_payload_RbKPgq/ansible_cisco.mso.mso_version_payload.zip/ansible/module_utils/connection.py", line 190, in __rpc__
    raise ConnectionError(to_text(msg, errors='surrogate_then_replace'), code=code)
fatal: [10.122.186.69]: FAILED! => {
    "ansible_facts": {
        "discovered_interpreter_python": "/usr/bin/python"
    }, 
    "changed": false, 
    "current": {}, 
    "invocation": {
        "module_args": {
            "host": "10.122.186.69", 
            "hostname": "10.122.186.69", 
            "login_domain": null, 
            "output_level": "debug", 
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "port": null, 
            "state": "query", 
            "timeout": 30, 
            "use_proxy": null, 
            "use_ssl": null, 
            "username": "admin", 
            "validate_certs": false
        }
    }, 
    "msg": "Authentication failed: Request failed: '_io.BytesIO' object has no attribute 'getbuffer'"
}
```

I have the same playbook working for a different user on the Ansible server

```
TASK [Get MSO version from MSO >= 3.2] *************************************************************************************************************************************************************************************************************************************************************************************************************************************************************
task path: /usr/abc1/playbooks/mso-adv-bootcamp_undeploy_schema_template_tenant_abc1_only.yml:43
<10.122.186.69> ESTABLISH LOCAL CONNECTION FOR USER: abc1
<10.122.186.69> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/abc1/.ansible/tmp/ansible-local-206985cm9Z5 `"&& mkdir "` echo /home/abc1/.ansible/tmp/ansible-local-206985cm9Z5/ansible-tmp-1677697389.79-20707-132724077259211 `" && echo ansible-tmp-1677697389.79-20707-132724077259211="` echo /home/abc1/.ansible/tmp/ansible-local-206985cm9Z5/ansible-tmp-1677697389.79-20707-132724077259211 `" ) && sleep 0'
<10.122.186.69> Attempting python interpreter discovery
<10.122.186.69> EXEC /bin/sh -c 'echo PLATFORM; uname; echo FOUND; command -v '"'"'/usr/bin/python'"'"'; command -v '"'"'python3.7'"'"'; command -v '"'"'python3.6'"'"'; command -v '"'"'python3.5'"'"'; command -v '"'"'python2.7'"'"'; command -v '"'"'python2.6'"'"'; command -v '"'"'/usr/libexec/platform-python'"'"'; command -v '"'"'/usr/bin/python3'"'"'; command -v '"'"'python'"'"'; echo ENDFOUND && sleep 0'
<10.122.186.69> EXEC /bin/sh -c '/usr/bin/python && sleep 0'
Using module file /home/abc1/.ansible/collections/ansible_collections/cisco/mso/plugins/modules/mso_version.py
<10.122.186.69> PUT /home/abc1/.ansible/tmp/ansible-local-206985cm9Z5/tmp9ncuUb TO /home/abc1/.ansible/tmp/ansible-local-206985cm9Z5/ansible-tmp-1677697389.79-20707-132724077259211/AnsiballZ_mso_version.py
<10.122.186.69> EXEC /bin/sh -c 'chmod u+x /home/abc1/.ansible/tmp/ansible-local-206985cm9Z5/ansible-tmp-1677697389.79-20707-132724077259211/ /home/abc1/.ansible/tmp/ansible-local-206985cm9Z5/ansible-tmp-1677697389.79-20707-132724077259211/AnsiballZ_mso_version.py && sleep 0'
<10.122.186.69> EXEC /bin/sh -c '/usr/bin/python /home/abc1/.ansible/tmp/ansible-local-206985cm9Z5/ansible-tmp-1677697389.79-20707-132724077259211/AnsiballZ_mso_version.py && sleep 0'
<10.122.186.69> EXEC /bin/sh -c 'rm -f -r /home/abc1/.ansible/tmp/ansible-local-206985cm9Z5/ansible-tmp-1677697389.79-20707-132724077259211/ > /dev/null 2>&1 && sleep 0'
ok: [10.122.186.69] => {
    "ansible_facts": {
        "discovered_interpreter_python": "/usr/bin/python"
    }, 
    "changed": false, 
    "current": {
        "id": "6139014f1c0000e600e7c786", 
        "platformVersion": "2.0.2h", 
        "schemaversion": "", 
        "timestamp": "2021-09-08T18:30:39.561279062", 
        "version": "3.4.1a"
    }, 
    "invocation": {
        "module_args": {
            "host": "10.122.186.69", 
            "hostname": "10.122.186.69", 
            "login_domain": null, 
            "output_level": "normal", 
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "port": null, 
            "state": "query", 
            "timeout": 30, 
            "use_proxy": null, 
            "use_ssl": null, 
            "username": "admin", 
            "validate_certs": false
        }
    }
}
Read vars_file '/usr/abc1/inventory/externalvars_ndo.yml'
```

* xxxx

### Affected Module Name(s):

* mso_XXXXX

### MSO version and MSO Platform

ND: "2.0.2h", 
NDO: "3.4.1a"

**Comment by akinross** [comment]
The getbuffer() method used on the '_io.BytesIO' object ( ansible-nd httpapi connection plugin ) does not exist in python3. Created a PR in ansible-nd to use python2.7 supported functions.

**Comment by lhercot** [comment]
Fixed by merging PR ansible-nd#41
